Added persistent score storage and gestion

f-michaut opened the pull request:
- Added .vsconfig to gitignore
- Created ScoreManager class
- Added functions to add new scores, list the current ones, and clear the scores
- At startup, load the score list from a file
- All the functions that change the score list write the new content to the disk

f-michaut committed:
Added persistent score storage and gestion

- Added .vsconfig to gitignore
- Created ScoreManager class
- Added functions to add new scores, list the current ones, and clear the scores
- At startup, load the score list from a file
- All the functions that change the score list write the new content to the disk

ClaudiaLapalme commented on Assets/Code/Scripts/ScoreManager.cs:
```diff
@@ -0,0 +1,80 @@
+public class ScoreManager : MonoBehaviour
+{
+    [Serializable]
+    public class ScoreData
```
I didn't realize that you could serialize a class. This is very cool and useful for testing

f-michaut replied:
You can serialize a class, and most of c#'s containers are also serializable, meaning you can serialize a class containing Lists or Disctionary of other classes. Altho, i think there is some restrictions, i don't remember them exactly tho

ClaudiaLapalme commented on Assets/Code/Scripts/ScoreManager.cs:
```diff
@@ -0,0 +1,80 @@
+    {
+        if (scores.Count < maxHighscores)
+            return true;
+        foreach (ScoreData score in scores) {
```
Wouldn't it be more efficient to just check the last value (or first, depending on if it's ascending or descending). I noticed that you already sort when a new high score is entered anyways.

f-michaut replied:
Oh you are right, i didn't think of taking adventage of the sorting 
I think i just need to check if the value is comprised between the first and last values
I'll change that

ClaudiaLapalme commented on Assets/Code/Scripts/ScoreManager.cs:
```diff
@@ -0,0 +1,80 @@
+        if (!IsHighScore(score))
+            return;
+        scores.Add(new ScoreData { date = System.DateTime.Now, name = name, score = score });
+        scores.Sort((ScoreData a, ScoreData b) => { return a.score - b.score; });
```
I'm a little confused by the return on this line. What does it do exactly?

f-michaut replied:
It's the return of the lambda function the Sort function will use to compare values in the list. It return an integer indicating which value is the biggest. So it return an integer less than, equal to or bigger than 0 following the same convention as in this page : https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.comparer-1.compare?view=net-5.0

ClaudiaLapalme reviewed (approved):
great work!

f-michaut committed:
#16 - More efficient IsHighScore function

- Replaced the for loop by a single check, since we only need to know
if the score is higher than the smaller one

tedgag reviewed (approved):
Looks good! Approved
